### PR TITLE
[FLINK-32084][checkpoint] Migrate current file merging of channel state snapshot into the unify file merging framework

### PR DESCRIPTION
**Pull request by fredia**
## What is the purpose of the change

Migrate current file merging of channel state snapshot into the unify file merging framework.
This PR only focuses on the snapshot/writing part.

## Brief change log
- Introduce `UnifyMergingChannelStateWriteRequestDispatcher` which construct a ` ChannelStateCheckpointWriter` that uses unified file merging mechanism to write channel state.
- Update `ChannelStateWriteRequestExecutorFactory#getOrCreateExecutor` , different ways to merge small files can be chosen here.

## Verifying this change

Please make sure both new and modified tests in this PR follow [the conventions for tests defined in our code quality guide](https://flink.apache.org/how-to-contribute/code-style-and-quality-common/#7-testing).

This change is already covered by existing tests, such as *(please describe tests)*.
- `ChannelStateWriteRequestDispatcherTest`
- `ChannelStateWriteRequestExecutorFactoryTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (checkpointing)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? ( no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by fredia** [comment]
@Zakelly @1996fanrui would you please to take a look? thanks!

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/UnifyMergingChannelStateWriteRequestDispatcher.java
Comment (1996fanrui) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/UnifyMergingChannelStateWriteRequestDispatcher.java:52]: The only difference between `UnifyMergingChannelStateWriteRequestDispatcher` and `ChannelStateWriteRequestDispatcherImpl` is how to `resolveCheckpointStorageLocation`, right?

If so, could we not introduce `UnifyMergingChannelStateWriteRequestDispatcher`? It means we put the whole logic into `ChannelStateWriteRequestDispatcherImpl`. We can pass `unifyFileMerging` and `CheckpointStorageAccess` to `ChannelStateWriteRequestDispatcherImpl`.

The reason is I guess we only needs `unified file merging mechanism` in the future, and `ChannelStateWriteRequestDispatcherImpl` can serve `unified file merging mechanism` directly.  

WDYT?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestExecutorFactory.java
Comment (1996fanrui) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestExecutorFactory.java:87]: nit: Could we new `DispatcherImpl` when `executor == null`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestExecutorFactory.java
Comment (1996fanrui) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestExecutorFactory.java:80]: IIUC, all subtasks of one job share same dispatcher means all subtasks will write same channel state file, right?

If so, all subtasks will share single thread, I'm afraid the performance isn't enough.

**Review by 1996fanrui — COMMENTED** [review]
Thanks @fredia for the contribution!

I left some comments, please take a look in your free time, and please correct me if anything is wrong, thanks~

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/UnifyMergingChannelStateWriteRequestDispatcher.java
Comment (fredia) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/UnifyMergingChannelStateWriteRequestDispatcher.java:52]: > The only difference between `UnifyMergingChannelStateWriteRequestDispatcher` and `ChannelStateWriteRequestDispatcherImpl` is how to `resolveCheckpointStorageLocation`, right?

Yes, it is.  

Good suggestion, I'll replace `checkpointStorage` with `checkpointStorageAccess` in  `ChannelStateWriteRequestDispatcherImpl`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestExecutorFactory.java
Comment (fredia) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestExecutorFactory.java:80]: Right, all subtasks will share single thread in this PR, good catch!

I plan to change it to a subtask using an executor, like things before FLINK-26803. Is this feasible?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java
Comment (1996fanrui) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:241]: If we hard code `maxSubtasksPerChannelStateFile` is 1, it means `execution.checkpointing.unaligned.max-subtasks-per-channel-state-file` doesn't take effect  after this PR, right?

I'm thinking could we keep the `execution.checkpointing.unaligned.max-subtasks-per-channel-state-file` first? and deprecate it after `unified file merging mechanism` is stable.

WDYT?

Comment (fredia) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:241]: > If we hard code maxSubtasksPerChannelStateFile is 1, it means execution.checkpointing.unaligned.max-subtasks-per-channel-state-file doesn't take effect after this PR, right?

Right.

In fact, both merge mechanisms can work simultaneously. The `unified file merging mechanism` allows different output streams to be written to a physical file, and the current merging mechanism allows multiple subtasks to be written to one output stream. 

I will keep `execution.checkpointing.unaligned.max-subtasks-per-channel-state-file` first.
- If `execution.checkpointing.unaligned.max-subtasks-per-channel-state-file:1` and `state.checkpoints.file-merging.enabled: true`, only the `unified file merging mechanism` takes effect.
- if `execution.checkpointing.unaligned.max-subtasks-per-channel-state-file > 1` and `state.checkpoints.file-merging.enabled: false`, only the current mechanism takes effect.
-  if `execution.checkpointing.unaligned.max-subtasks-per-channel-state-file > 1` and `state.checkpoints.file-merging.enabled: true`, both two mechanism take effect.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java
Comment (1996fanrui) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java:338]: Wow, this comment is useful for understanding, great job.

**Review by 1996fanrui — APPROVED** [review]
Thanks @fredia for the updated!

LGTM assuming the CI is green.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestDispatcherImpl.java
Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestDispatcherImpl.java:208]: Hi @fredia , it's still protected, is there any reason for it?

Comment (fredia) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestDispatcherImpl.java:208]: Aha, I forgot this, I will change it and squash commits.

**Comment by fredia** [comment]
When passing `CheckpointStorageWorkerView` directly to `ChannelStateWriteRequestDispatcher`,  `checkpointStorage` `UnalignedCheckpointRescaleITCase` might throw exceptions when the task exited.

This is mainly because the task will close the stream held by the current thread when it exits, see `FileSystemSafetyNet#closeSafetyNetAndGuardedResourcesForThread`(Thanks for @1996fanrui's help).

Therefore, I let  `ChannelStateWriteRequestDispatcher#CheckpointStorageWorkerView` be lazily initialized.
and changed the `filesystem` in `FsMergingCheckpointStorageAccess` to unwrapped `filesystem`, to prevent different threads from interfering with each other when exiting.

```
Caused by: java.io.IOException: Could not open output stream for state backend
	at org.apache.flink.runtime.state.filesystem.FsCheckpointStreamFactory$FsCheckpointStateOutputStream.createStream(FsCheckpointStreamFactory.java:461) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.state.filesystem.FsCheckpointStreamFactory$FsCheckpointStateOutputStream.flushToFile(FsCheckpointStreamFactory.java:308) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.state.filesystem.FsCheckpointStreamFactory$FsCheckpointStateOutputStream.write(FsCheckpointStreamFactory.java:284) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at java.io.DataOutputStream.write(DataOutputStream.java:107) ~[?:1.8.0_292]
	at java.io.FilterOutputStream.write(FilterOutputStream.java:97) ~[?:1.8.0_292]
	at org.apache.flink.runtime.io.network.buffer.NetworkBuffer.getBytes(NetworkBuffer.java:404) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateSerializerImpl.writeData(ChannelStateSerializer.java:164) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateCheckpointWriter.lambda$write$2(ChannelStateCheckpointWriter.java:194) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateCheckpointWriter.runWithChecks(ChannelStateCheckpointWriter.java:274) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateCheckpointWriter.write(ChannelStateCheckpointWriter.java:189) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateCheckpointWriter.writeInput(ChannelStateCheckpointWriter.java:153) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequest.lambda$write$2(ChannelStateWriteRequest.java:139) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequest.lambda$buildWriteRequest$8(ChannelStateWriteRequest.java:231) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.CheckpointInProgressRequest.execute(ChannelStateWriteRequest.java:366) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequestDispatcherImpl.handleCheckpointInProgressRequest(ChannelStateWriteRequestDispatcherImpl.java:169) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequestDispatcherImpl.dispatchInternal(ChannelStateWriteRequestDispatcherImpl.java:119) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequestDispatcherImpl.dispatch(ChannelStateWriteRequestDispatcherImpl.java:86) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequestExecutorImpl.loop(ChannelStateWriteRequestExecutorImpl.java:182) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequestExecutorImpl.run(ChannelStateWriteRequestExecutorImpl.java:136) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	... 1 more
Caused by: java.io.IOException: Cannot register Closeable, registry is already closed. Closing argument.
	at org.apache.flink.util.AbstractAutoCloseableRegistry.registerCloseable(AbstractAutoCloseableRegistry.java:89) ~[flink-core-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.core.fs.ClosingFSDataOutputStream.wrapSafe(ClosingFSDataOutputStream.java:103) ~[flink-core-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.core.fs.SafetyNetWrapperFileSystem.create(SafetyNetWrapperFileSystem.java:131) ~[flink-core-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.core.fs.EntropyInjector.createEntropyAware(EntropyInjector.java:76) ~[flink-core-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.state.filesystem.FsCheckpointStreamFactory$FsCheckpointStateOutputStream.createStream(FsCheckpointStreamFactory.java:451) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.state.filesystem.FsCheckpointStreamFactory$FsCheckpointStateOutputStream.flushToFile(FsCheckpointStreamFactory.java:308) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.state.filesystem.FsCheckpointStreamFactory$FsCheckpointStateOutputStream.write(FsCheckpointStreamFactory.java:284) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at java.io.DataOutputStream.write(DataOutputStream.java:107) ~[?:1.8.0_292]
	at java.io.FilterOutputStream.write(FilterOutputStream.java:97) ~[?:1.8.0_292]
	at org.apache.flink.runtime.io.network.buffer.NetworkBuffer.getBytes(NetworkBuffer.java:404) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateSerializerImpl.writeData(ChannelStateSerializer.java:164) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateCheckpointWriter.lambda$write$2(ChannelStateCheckpointWriter.java:194) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateCheckpointWriter.runWithChecks(ChannelStateCheckpointWriter.java:274) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateCheckpointWriter.write(ChannelStateCheckpointWriter.java:189) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateCheckpointWriter.writeInput(ChannelStateCheckpointWriter.java:153) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequest.lambda$write$2(ChannelStateWriteRequest.java:139) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequest.lambda$buildWriteRequest$8(ChannelStateWriteRequest.java:231) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.CheckpointInProgressRequest.execute(ChannelStateWriteRequest.java:366) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequestDispatcherImpl.handleCheckpointInProgressRequest(ChannelStateWriteRequestDispatcherImpl.java:169) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequestDispatcherImpl.dispatchInternal(ChannelStateWriteRequestDispatcherImpl.java:119) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequestDispatcherImpl.dispatch(ChannelStateWriteRequestDispatcherImpl.java:86) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequestExecutorImpl.loop(ChannelStateWriteRequestExecutorImpl.java:182) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
	at org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequestExecutorImpl.run(ChannelStateWriteRequestExecutorImpl.java:136) ~[flink-runtime-1.20-SNAPSHOT.jar:1.20-SNAPSHOT]
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java
Comment (1996fanrui) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:499]: Would you mind adding some comments here to explain why we don't pass `CheckpointStorageAccess` directly?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStorageAccess.java
Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStorageAccess.java:237]: Also, I have a concern here: When we use `GuardedFileSystem`, all streams can be closed properly when the corresponding thread is exiting. After this change, if we use `UnguardedFileSystem`, it may cause stream leak.

Comment (fredia) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStorageAccess.java:237]: > I'm thinking could checkpointsDirectory as the parameter of FsMergingCheckpointStorageAccess constructor?

👍 Agreed, I will update the constructor as your suggestion.

> Also, I have a concern here: When we use GuardedFileSystem, all streams can be closed properly when the corresponding thread is exiting. After this change, if we use UnguardedFileSystem, it may cause stream leak.

`FileMergingSnapshotManager`  takes this responsibility, `FileMergingSnapshotManager` would close all held physical files and streams when the task executor close/shutdown.

Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStorageAccess.java:237]: > FileMergingSnapshotManager takes this responsibility, FileMergingSnapshotManager would close all held physical files and streams when the task executor close/shutdown.

Sorry, I'm not familiar with `FileMergingSnapshotManager`, thanks for the clarification. Sounds good to me.

**Review by 1996fanrui — APPROVED** [review]

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java
Comment (Zakelly) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:498]:
```suggestion
                                    // when thread exits. Channel state writers are invoked in other
```

**Comment by Zakelly** [comment]
> When passing `CheckpointStorageWorkerView` directly to `ChannelStateWriteRequestDispatcher`, `UnalignedCheckpointRescaleITCase` might throw exceptions when the task exits.
> 
> This is mainly because the task will close the stream held by the current thread when it exits, see `FileSystemSafetyNet#closeSafetyNetAndGuardedResourcesForThread`(Thanks for @1996fanrui's help).
> 
> Therefore, I let `ChannelStateWriteRequestDispatcher#CheckpointStorageWorkerView` be lazily initialized. and changed the `filesystem` in `FsMergingCheckpointStorageAccess` to unwrapped `filesystem`, to prevent different threads from interfering with each other when exiting.

Will the `PhysicalFilePool` close while some streams are still being written? There might be leaks of opened streams?

**Comment by fredia** [comment]
> Will the `PhysicalFilePool` close while some streams are still being written? There might be leaks of opened streams?

`PhysicalFilePool` is managed at task manager level, and asynchronous checkpoint threads is managed by `StreamTask`, theoretically, it is closed after the asynchronous checkpoint thread exits, no stream is writing when `PhysicalFilePool` close.

**Comment by Zakelly** [comment]
> > Will the `PhysicalFilePool` close while some streams are still being written? There might be leaks of opened streams?
> 
> `PhysicalFilePool` is managed at task manager level, and asynchronous checkpoint threads is managed by `StreamTask`, theoretically, it is closed after the asynchronous checkpoint thread exits, no stream is writing when `PhysicalFilePool` close.

Emmmmm... I'm afraid the pool share the same lifecycle with `FileMergingSnapshotManager`, which is at `<job, taskmanager>` level. But I guess you are right, those job-level resources within TM are released after stream task quit.

Thanks for the explanation. I went through the checkpoint and restore code path, and have no further comments

**Review by Zakelly — APPROVED** [review]

**Pull request closed without merging**
